### PR TITLE
Feature/Add message read indicator

### PR DESCRIPTION
**Pull request by Milena-Czierlinski**
## Summary by CodeRabbit

- **New Features**
	- Introduced tracking for when messages are read with a new `wasReadAt` property.
	- Added the ability to mark messages as read or unread.
- **Refactor**
	- Reordered constructor parameters and swapped order of imports in certain classes for optimization.
- **Tests**
	- Implemented new tests for the read/unread functionality of messages.

### CODE REVIEW

**Comment by Milena-Czierlinski** [comment]
@coderabbitai review

File: packages/transport/test/modules/messages/MessageController.test.ts [hunk 1]
```diff
@@ -141,4 +141,33 @@ describe("MessageController", function () {
...
+        expect(message.wasReadAt!.isSameOrAfter(timeBeforeRead)).toBe(true);
+        expect(message.wasReadAt!.isSameOrBefore(timeAfterRead)).toBe(true);
+    });
+
+    test("should not change wasReadAt of a read message", async function () {
+        let message = await recipient.messages.getMessage(tempId1);
+        message = await recipient.messages.markMessageAsRead(message!.id);
+        const firstReadAt = message.wasReadAt;
+
+        message = await recipient.messages.markMessageAsRead(message.id);
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/messages/MessageController.test.ts:161]: same in the tests above

File: packages/transport/test/modules/messages/MessageController.test.ts [hunk 2]
```diff
@@ -141,4 +141,33 @@ describe("MessageController", function () {
         const messages = await recipient.messages.getMessagesByRelationshipId(relationshipId);
         expect(messages).toHaveLength(3);
     });
+
+    test("should mark an unread message as read", async function () {
+        let message = await recipient.messages.getMessage(tempId1);
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/messages/MessageController.test.ts]: great!

File: packages/transport/src/modules/messages/MessageController.ts [hunk 3]
```diff
@@ -204,6 +204,34 @@ export class MessageController extends TransportController {
...
+        const messageDoc = await this.messages.read(id.toString());
+        if (!messageDoc) {
+            throw CoreErrors.general.recordNotFound(Message, id.toString());
+        }
+
+        const message = Message.from(messageDoc);
+        if (typeof message.wasReadAt === "undefined") {
+            message.wasReadAt = CoreDate.utc();
+            await this.messages.update(messageDoc, message);
+        }
```
Comment (jkoenig134) [inline packages/transport/src/modules/messages/MessageController.ts:217]: please use early return where possible:

```suggestion
        if (typeof message.wasReadAt !== "undefined") return message;
        
        message.wasReadAt = CoreDate.utc();
        await this.messages.update(messageDoc, message);
```

File: packages/transport/src/modules/messages/MessageController.ts [hunk 4]
```diff
@@ -204,6 +204,34 @@ export class MessageController extends TransportController {
...
+
+    public async markMessageAsUnread(id: CoreId): Promise<Message> {
+        const messageDoc = await this.messages.read(id.toString());
+        if (!messageDoc) {
+            throw CoreErrors.general.recordNotFound(Message, id.toString());
+        }
+
+        const message = Message.from(messageDoc);
+        message.wasReadAt = undefined;
+        await this.messages.update(messageDoc, message);
```
Comment (jkoenig134) [inline packages/transport/src/modules/messages/MessageController.ts:230]: you can skip a db call by returning when wasReadAt was already undefined

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

File: packages/transport/test/modules/messages/MessageController.test.ts [hunk 5]
```diff
@@ -141,4 +141,33 @@ describe("MessageController", function () {
...
+        const timeBeforeRead = CoreDate.utc();
+        message = await recipient.messages.markMessageAsRead(message!.id);
+        const timeAfterRead = CoreDate.utc();
+
+        expect(message.wasReadAt).toBeDefined();
+        expect(message.wasReadAt!.isSameOrAfter(timeBeforeRead)).toBe(true);
+        expect(message.wasReadAt!.isSameOrBefore(timeAfterRead)).toBe(true);
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/messages/MessageController.test.ts]: I looked into the impl. Looks like isBetween includes start, but excludes end. You could add one millisecond to the end time to get rid of this.

File: packages/runtime/test/transport/messages.test.ts [hunk 6]
```diff
@@ -1,27 +1,35 @@
...
 let eventBus1: MockEventBus;
 let transportServices2: TransportServices;
+let client1: TestRuntimeServices;
+let client2: TestRuntimeServices;
 
 beforeAll(async () => {
     const runtimeServices = await serviceProvider.launch(2);
+    client1 = runtimeServices[0];
+    client2 = runtimeServices[1];
+    await ensureActiveRelationship(client1.transport, client2.transport);
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/runtime/test/transport/messages.test.ts:28]: I see, I'm okay with both. I was just curious :)

File: packages/transport/test/modules/messages/MessageController.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/messages/MessageController.test.ts]: But I think we have to put the whole transport and consumption libs tests on that TODO list :D

**Review by jkoenig134 — APPROVED** [review]

File: packages/transport/src/modules/messages/MessageController.ts (hunk 3 above)
Comment (jkoenig134) [reply packages/transport/src/modules/messages/MessageController.ts:217]: this was marked as resolved, but not resolved

File: packages/transport/src/modules/messages/MessageController.ts (hunk 4 above)
Comment (jkoenig134) [reply packages/transport/src/modules/messages/MessageController.ts:230]: same here.

**Review by jkoenig134 — CHANGES_REQUESTED** [review]
some still open comments

**Review by jkoenig134 — APPROVED** [review]

**Pull request merged**